[CMSP-992] UI changes

jazzsequence opened the pull request:
This PR:

* updates some of the language in the notices and alerts
* adds a new build pipeline using Gulp and Sass (and removes old NPM dependencies that weren't being used)
* adds `build_node_assets` to the build-tag-release workflow
* adds styling for Pantheon Page Cache Max Age field (uses selectors added in https://github.com/pantheon-systems/pantheon-mu-plugin/pull/46)
* adds new elements and overrides default HTML elements (based on hooks added in https://github.com/pantheon-systems/pantheon-mu-plugin/pull/46)
* adds styles (and a filter that is commented-out) for an info bar in the Max Age setting per the Figma designs (disabled because we aren't using it)
* adds other styles and text per the Figma designs
* changes the input box to a select (dropdown) if the value is not filtered and not manually set. (a warning is displayed if a custom value is saved that is not one of the recommended (default) options
* fixes some logic around warnings and alerts
* adds tests for new components

## Demo
https://www.loom.com/share/686250ad45b3493fa5f8f250cf233345

## Screenshots
**Global notice when max age of 600 is reset to the default**
<img width="1132" alt="Screenshot 2024-05-20 at 2 37 35 PM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/5bbe56dc-bf32-49fc-b51a-82a15ce4d76d">

**Pantheon Page Cache warning when max age is _far_ less than the recommended value** (Note: `max-age` is now `max age` in the code)
<img width="1062" alt="Screenshot 2024-05-20 at 2 37 28 PM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/4bd0ef7d-9113-43f8-99bc-d9374be2f7dc">

**Site Health recommendation when max age is _far_ less than the recommended value** (Note: all instances of `max-age`/`Max-Age` have been replaced with their equivalent without the `-` in the code)
<img width="1558" alt="Screenshot 2024-05-21 at 9 14 44 AM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/1eebb5ad-bf12-4c0e-938a-676607ecb579">

**Pantheon Page Cache warning when max age is closer, but still less than, the recommended value** (Note: `max-age` is now `max age` in the code)
<img width="719" alt="Screenshot 2024-05-20 at 2 37 07 PM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/1c12b8ee-978a-4761-b332-a246edc63cd3">

**Site Health recommendation when max age is closer, but still less than, the recommended value** (Note: all instances of `max-age`/`Max-Age` have been replaced with their equivalent without the `-` in the code)
<img width="1563" alt="Screenshot 2024-05-21 at 9 11 43 AM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/eb006d7c-e539-4745-b9f4-ab874c9b58ac">

**Pantheon Page Cache options page with new settings**
<img width="1563" alt="Screenshot 2024-05-21 at 9 09 00 AM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/89392d79-55ba-4a4e-bdb3-9d82a124837e">

**Pantheon Page Cache Max Age dropdown**
<img width="1555" alt="Screenshot 2024-05-21 at 9 09 15 AM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/cb162019-5d74-476c-a2df-9dd573e5c2ba">

**Pantheon Page Cache Max Age with custom value** (Note: This would only occur if a value other than 600 and any of the default options had been set previously.)
<img width="1552" alt="Screenshot 2024-05-21 at 1 59 30 PM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/851da312-bdac-4f92-b0b1-aae06b249fce">

**Pantheon Page Cache Max Age with _filtered_ value** (Note: If a custom/different value is saved in the database, the filtered value replaces it in the (disabled) input)
<img width="1559" alt="Screenshot 2024-05-21 at 3 28 41 PM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/da07a0e5-e3dc-4aec-90cb-91d887aed9bd">

**Pantheon Cache Max Age with info bar** (Note: This is controlled via a filter that can be hooked to `__return_true`/`__return_false` to toggle. By default it is not displayed. Styles are included in this PR per the Figma and in case we choose to use it in the future.
<img width="1561" alt="Screenshot 2024-05-21 at 3 30 20 PM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/f0d73d0a-e217-4482-91a3-63e4a232b3cd">

jazzsequence committed:
add initial admin interface class
hook to mu-plugin > 1.4.0

jazzsequence committed:
some more scaffolding

jazzsequence committed:
require at least 6.4

jazzsequence committed:
don't show notices if wp_admin_notice doesn't exist

jazzsequence committed:
add notice if mu-plugin was not found

jazzsequence committed:
add notice if the mu-plugin is old

jazzsequence committed:
fix link and change decimal to string

jazzsequence committed:
use spaces instead of line breaks

jazzsequence committed:
add helper function to get current max age
this is either set in the mu-plugin or empty. if empty, it should default to 1 week

jazzsequence committed:
add notice if less than 1 week but not == 600 seconds

jazzsequence committed:
add docblock

jazzsequence committed:
add translator comments

jazzsequence committed:
update tested up to and requires at least values in readmes

jazzsequence committed:
version bump

jazzsequence committed:
bump composer deps

jazzsequence committed:
don't display the warnings if we aren't on the pantheon cache admin page

jazzsequence committed:
convert static class to a namespace
there's no reason to make this a class since we're not using any class features

jazzsequence committed:
add a bootstrapper for namespace-only files

jazzsequence committed:
lint

jazzsequence committed:
escape the wp_die message

jazzsequence committed:
use a helper for humanized time calculations

jazzsequence committed:
we still need to test the current max age

jazzsequence committed:
pass the current max age to the message, too
and be specific that we're talking about seconds

jazzsequence committed:
use the filterable value for default max age
because it could be something different

jazzsequence committed:
add a helper function for retrieving the default max age
this is helpful so we can always return the filtered value

jazzsequence committed:
also display the seconds in the successful test

jazzsequence committed:
rank the max age value
don't say it's "very low" unless that's actually accurate.

jazzsequence committed:
add caching to the ranked compare

jazzsequence committed:
make the color more severe if it's a bigger issue

jazzsequence committed:
change the color of the admin notice based on the severity of the diff

jazzsequence committed:
clear the transient when the option is updated

jazzsequence committed:
break notice type into a variable

jazzsequence committed:
fix the translator note

jazzsequence committed:
keep the rank but hide it from the dom

jazzsequence committed:
add unit tests

jazzsequence committed:
add initial behat test
let's just see if we can assume this will work

jazzsequence committed:
fix failing test

jazzsequence committed:
specify the element where the notices are

jazzsequence committed:
add site health check test

jazzsequence committed:
actually save the changes

jazzsequence committed:
remove max age check
only exists in an html comment anyway

jazzsequence committed:
simplify the site health check

jazzsequence committed:
add the message when you expand the accordion

jazzsequence committed:
add other site health things

jazzsequence committed:
change press to click?

jazzsequence committed:
add "open the accordion" as a custom step

jazzsequence committed:
move behat stuff around

jazzsequence committed:
move stuff around again

jazzsequence committed:
update the behat.yml

jazzsequence committed:
remove custom context and try "when I follow"

jazzsequence committed:
move to checking the text inside the elements

jazzsequence committed:
target the hidden element

jazzsequence committed:
use hidden='hidden'

jazzsequence committed:
call it a div

jazzsequence committed:
simplify the tests

jazzsequence committed:
remove press
since we can't actually press

jazzsequence committed:
it did actually find the right heading
maybe it will find the right text?

jazzsequence committed:
add @\since tags

jazzsequence committed:
add changelog and upgrade notice

jazzsequence committed:
let's try checking for some of the hidden text

jazzsequence committed:
period at the end of the sentence

jazzsequence committed:
add the ranks back
this shows us that our ranking function is working

jazzsequence committed:
add the messages

jazzsequence committed:
remove age rank tests
because goutte doesn't natively interpret html comments

jazzsequence committed:
add github testing badge
since phpunit is no longer run on circle

jazzsequence committed:
add script to maybe skip behat tests if we don't need to run them

jazzsequence committed:
add step to workflow

jazzsequence committed:
do not run if the circle config changes

jazzsequence committed:
also ignore if scripts (that aren't behat-related) have changed

jazzsequence committed:
fix the logic (again)

jazzsequence committed:
one more flippity flip

jazzsequence committed:
remove the "ignoring" message
it's actually implied by the fact that is_ignored is checked

jazzsequence committed:
use function to determine if a file should be ignored

Co-authored-by: Phil Tyler <phil.tyler@pantheon.io>

jazzsequence committed:
update changelog to note mu-plugin updates

jazzsequence committed:
add functionality to automatically set the default_ttl if it was set to 600

jazzsequence committed:
trigger the setting change and display an admin notice informing users it was done
only show the notice to users who haven't already seen it and who can manage_options

jazzsequence committed:
add unit tests for set_max_age_to_default

jazzsequence committed:
move the max age update notice to the bootstrap

jazzsequence committed:
remove the set_max_age_updated helper
this wasn't working and meant that we were needing to make multiple update options in a short space of time. we might have actually been running into race conditions where one update was overwriting the other one.

jazzsequence committed:
use a new option for the updated flag

jazzsequence committed:
add _notice to the user meta key so it's distinct from the new option

jazzsequence committed:
update tests for new changes

jazzsequence committed:
add translator note

jazzsequence committed:
user notice about 600 second update is not dismissable
because it only appears once

jazzsequence committed:
don't show seconds in updated notice

jazzsequence committed:
pass the function into the apply_filter hook
so you can selectively disable specific notices

jazzsequence committed:
add docblock

jazzsequence committed:
add test for the notice

jazzsequence committed:
return the value of disable_notices
rather than hard-coded "false"

jazzsequence committed:
Add scenario to set max-age to 600 and auto-update to the default value

jazzsequence committed:
here's a behat test that might not work
not sure if it will know what pantheon-cache[default-ttl] is without giving it an element to look at.

jazzsequence committed:
fixed it based on upstream tests

jazzsequence committed:
refresh the page

jazzsequence committed:
go to instead of refresh
since refresh isn't a thing ¯\_(ツ)_/¯

jazzsequence committed:
remove all the old npm stuff and use all new npm stuff

jazzsequence committed:
add node build to the build & tag workflow

jazzsequence committed:
add del package

jazzsequence committed:
repurpos gruntfile into gulpfile

jazzsequence committed:
repackage gulpfile and package.json as a module

jazzsequence committed:
add additional test failure messaging

jazzsequence committed:
Merge branch 'cmsp-994' into cmsp-992-ui

jazzsequence committed:
gitignore assets

jazzsequence committed:
don't export the gulpfile

jazzsequence committed:
invoke sass correctly

jazzsequence committed:
proper sassing

jazzsequence committed:
add a main.scss file

jazzsequence committed:
Merge branch 'develop' into cmsp-992-ui

jazzsequence committed:
compiler tweaks

jazzsequence committed:
use concat

jazzsequence committed:
initial styles

jazzsequence committed:
enqueue css

jazzsequence committed:
use the filter added in https://github.com/pantheon-systems/pantheon-mu-plugin/pull/46 to add header

jazzsequence committed:
some initial styles for the max age box

jazzsequence committed:
add a global admin notice that appears once

jazzsequence committed:
remove seconds from the notice

jazzsequence committed:
test the new notice

jazzsequence committed:
set dismissable depending on which notice we're showing

jazzsequence committed:
show the max age rank html comment

jazzsequence committed:
fix the conditionals for cs

jazzsequence committed:
update behat tests for new notice language

jazzsequence committed:
bugfix: don't update user meta that max_age was updated if it hasn't been

jazzsequence committed:
add one more thing to test in behat

jazzsequence committed:
add a test for the error, specifically

jazzsequence committed:
add a test for the warning

jazzsequence committed:
simplify the conditionals

jazzsequence committed:
lint

jazzsequence committed:
specify the pages that we're on?

jazzsequence committed:
remove specific element

jazzsequence committed:
fix typo

jazzsequence committed:
take out the dashboard main page goto

jazzsequence committed:
validate the setting was saved

jazzsequence committed:
let's just remove the 5 days notice

jazzsequence committed:
just remove the test
we have similar things being tested in phpunit

jazzsequence committed:
we can actually remove current_max_age >= WEEK_IN_SECONDS
this is covered by max_age_rank. if max_age_rank is 0 it, by definition, means that the current max age is >= WEEK_IN_SECONDS

jazzsequence committed:
if we're only comparing two things, we don't need to break the if clauses up

jazzsequence committed:
we do want to warn users if they have manually set their max age to 600
we just don't want to warn them if we're about to (or just did) change it to a week for them

jazzsequence committed:
Merge branch 'cmsp-995-one-more-notice' into cmsp-992-ui

jazzsequence committed:
there should only be one css file

jazzsequence committed:
remove seconds

jazzsequence committed:
check stored default_ttl value

jazzsequence committed:
fix failing test

jazzsequence committed:
fix translators note

jazzsequence committed:
remove seconds

jazzsequence committed:
remove redundant check

jazzsequence committed:
tweak the logic again

jazzsequence committed:
fix the test by updating the default ttl

jazzsequence committed:
add the description after the input

jazzsequence committed:
styles for box and info bar

jazzsequence committed:
update the filter name

jazzsequence committed:
filter the input box

jazzsequence committed:
add a padding helper

jazzsequence committed:
...and use it

jazzsequence committed:
lint

jazzsequence committed:
Merge branch 'develop' into cmsp-992-ui

jazzsequence committed:
add styling for links in the info bar.

jazzsequence committed:
no padding on the right
padding the right makes the bar flush with the edge of the box

jazzsequence committed:
adjust h3 margin
so it looks better when there's no banner

jazzsequence committed:
style the select input

jazzsequence committed:
don't render the info bar

jazzsequence committed:
use wp_kses_post instead of esc_html_e
we might want to add links to this bar

jazzsequence committed:
helper function for available dropdown options and make it filterable

jazzsequence committed:
Refactor default_ttl_input function to handle custom options and filterable dropdown

jazzsequence committed:
remove seconds from behat tests

jazzsequence committed:
add a period for dumb lint rule

jazzsequence committed:
remove hyphen from max-age
in everything user-facing

jazzsequence committed:
update the tests

jazzsequence committed:
conditionally append a timestamp to the css version
to refresh it on each page load if debug is true

jazzsequence committed:
don't show the input box if the ttl is unset

jazzsequence committed:
remove filters in teardown

jazzsequence committed:
remove hyphen in tests

jazzsequence committed:
specify method visibility

jazzsequence committed:
test the setting description callback

jazzsequence committed:
test the default ttl option input callback

jazzsequence committed:
test max_age_options and filter

jazzsequence committed:
add a behat test for the select input

jazzsequence committed:
maybe change the then

jazzsequence committed:
remove all hyphens

jazzsequence committed:
be more specific about the select input

jazzsequence committed:
try using the class and the select

jazzsequence committed:
let's try checking to see if the response contains the selected option

jazzsequence committed:
print for debugging

jazzsequence committed:
one last try before I give up

jazzsequence committed:
remove the new behat test

jazzsequence committed:
use a simple true/false gate for max_age header

jazzsequence committed:
display a warning if the default_ttl is saved to a value that's not one of the default optiosn

jazzsequence committed:
show the dropdown with the custom value
so a new value can be selected from the droptdown when prompted

jazzsequence committed:
remove placeholder comment

jazzsequence committed:
gix the global warning shown logic

jazzsequence committed:
make sure to take into consideration custom and filtered

jazzsequence committed:
bail early if the max age is filtered

jazzsequence committed:
better comment

pwtyler commented on .gitattributes:
```diff
@@ -19,9 +19,9 @@
 /composer.lock export-ignore
 /CONTRIBUTING.md export-ignore
-/Gruntfile.js export-ignore
+/gulpfile.mjs export-ignore
```
confirming, this should be a replacement, not an adendum?

jazzsequence replied:
Correct. Using Gulp _instead of_ Grunt (which we weren't really using for anything anyway).

scottbuscemi commented:
<img width="307" alt="Screenshot 2024-05-21 at 3 00 43 PM" src="https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/4336848/647f04ff-b965-4254-bf04-86066b13dd4b">

Looks like "Recommended Setting" is using the hardcoded value rather than just saying '1 day'

jazzsequence commented:
@scottbuscemi 
> Looks like "Recommended Setting" is using the hardcoded value rather than just saying '1 day'

Explained in the demo. If a value is filtered, the filtered value takes the place of the recommendation/becomes the new recommendation. If this is different than what you would like, we can change it, but the assumption is, if you're filtering it, you're breaking the seal on the warranty.

scottbuscemi commented:
Let's remove the "Value range" sentence

jazzsequence committed:
remove value range entirely

scottbuscemi commented:
If someone is doing their own thing, we can just not recommend something anymore :)

jazzsequence commented:
> If someone is doing their own thing, we can just not recommend something anymore :)

Sure can.

pwtyler commented on inc/admin-interface.php:
```diff
@@ -36,6 +35,149 @@ function bootstrap() {
+		return;
+	}
+
+	$debug = defined( 'WP_DEBUG' ) && WP_DEBUG ? '-' . time() : '';
+
+	wp_enqueue_style( 'papc-admin', plugin_dir_url( __DIR__ ) . 'assets/css/styles.css', [], '2.0.0' . $debug ); // If WP_DEBUG is true, append a timestamp to the end of the path so we get a fresh copy of the css.
```
```suggestion
	// If WP_DEBUG is true, append a timestamp to the end of the path so we get a fresh copy of the css.
	$debug = defined( 'WP_DEBUG' ) && WP_DEBUG ? '-' . time() : '';
	wp_enqueue_style( 'papc-admin', plugin_dir_url( __DIR__ ) . 'assets/css/styles.css', [], '2.0.0' . $debug );
```

jazzsequence committed:
remove recommendation if filtered

jazzsequence commented:
![Screenshot 2024-05-21 at 4 07 26 PM](https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/a9248f2f-da69-4a0d-8750-68c708fd79b1)

@scottbuscemi

pwtyler commented on inc/admin-interface.php:
```diff
@@ -36,6 +35,149 @@ function bootstrap() {
+	wp_enqueue_style( 'papc-admin', plugin_dir_url( __DIR__ ) . 'assets/css/styles.css', [], '2.0.0' . $debug ); // If WP_DEBUG is true, append a timestamp to the end of the path so we get a fresh copy of the css.
+}
+
+/**
+ * Add a header to the max-age setting field.
+ */
```
missing since and return

pwtyler commented on inc/admin-interface.php:
```diff
@@ -36,6 +35,149 @@ function bootstrap() {
+	return ob_get_clean();
+}
+
+/**
+ * Add a description to the max-age setting field.
+ */
```
since and return 🔢

pwtyler reviewed (commented):
A few comments. have not watched recording yet.

jazzsequence committed:
break out custom and filtered outputs
filtered displays the grayed out input box
custom adds to the dropdown

jazzsequence committed:
update tests

jazzsequence committed:
translator note

jazzsequence committed:
move the debug note

Co-authored-by: Phil Tyler <phil.tyler@pantheon.io>

jazzsequence committed:
add since and return values

jazzsequence commented:
Cache max age recommendation message updates:

**Current value is _above_ the recommendation**
![Screenshot 2024-05-22 at 9 05 20 AM](https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/17c03429-b179-43c4-9751-032c7e671803)

**Current value is set to the recommendation**
![Screenshot 2024-05-22 at 9 05 38 AM](https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/736789a4-5efe-49b0-b461-dcd31ea08680)

**Current value is _below_ the recommendation**
![Screenshot 2024-05-22 at 9 06 08 AM](https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/1a728b56-6fb9-4e36-9e44-ef8c5c14a708)

**Current value is _filtered below_ the recommendation**
![Screenshot 2024-05-22 at 9 06 37 AM](https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/71576932-1b41-4d92-b06a-7580c744ff3d)

**Current value is _filtered above_ the recommendation**
![Screenshot 2024-05-22 at 9 07 05 AM](https://github.com/pantheon-systems/pantheon-advanced-page-cache/assets/991511/8df7820c-6f89-4d35-bbeb-3092e8af44d1)

**Note:** Updated to remove duplicative "current" in the message(s).

jazzsequence committed:
push the description down a bit

jazzsequence committed:
personalized message based on current cache max age

jazzsequence committed:
remove current when not necessary

jazzsequence committed:
update tests, add a data provider

pwtyler commented on inc/admin-interface.php:
```diff
@@ -36,6 +35,160 @@ function bootstrap() {
+	$below_recommended_message = __( 'Your cache maximum age is currently <strong>below</strong> the recommended value.', 'pantheon-advanced-page-cache' );
+	$recommended_message = __( 'Your cache maximum age is currently set to the recommended value.', 'pantheon-advanced-page-cache' );
+	$recommendation_message = get_current_max_age() > WEEK_IN_SECONDS ? $above_recommended_message : ( get_current_max_age() < WEEK_IN_SECONDS ? $below_recommended_message : $recommended_message );
+	$filtered_message = $filtered ? sprintf(
+		// translators: %s is the humanized max-age.
+		__( 'This value has been hardcoded to %s via a filter.', 'pantheon-advanced-page-cache' ),
+		'<strong>' . humanized_max_age() . '</strong>'
+	) : '';
+	$pantheon_cache = get_option( 'pantheon-cache', [] );
+	$custom_ttl = isset( $pantheon_cache['default_ttl'] ) && ! array_key_exists( $pantheon_cache['default_ttl'], max_age_options() );
+	$filtered_message .= $custom_ttl && ! $filtered ? '<br />' . __( '<strong>Warning:</strong>The cache max age is not one of the recommended values. If this is not intentional, you should remove this custom value and save the settings, then select one of the options from the dropdown.', 'pantheon-advanced-page-cache' ) : '';
```
This is dense to read and parse out reading code but not going to block on it. For example, if $custom_ttl is a bool, the variable name should read as one.

https://github.com/pantheon-systems/Standards/blob/b0a15d37e61461226f1d6061bb3ac5d087548948/Data_Standards.md?plain=1#L15-L16

I know you love ternary operators but it's a pain long-term.

pwtyler commented on inc/admin-interface.php:
```diff
@@ -36,6 +35,160 @@ function bootstrap() {
+ * @return int
+ */
+function update_default_ttl_input( $default_input ) {
+	$slug = 'pantheon-cache';
+	$pantheon_cache = get_option( $slug, [] );
+	$default_ttl = isset( $pantheon_cache['default_ttl'] ) && $pantheon_cache['default_ttl'] !== 0 ? $pantheon_cache['default_ttl'] : WEEK_IN_SECONDS;
+	$options = max_age_options();
+	$filtered = has_filter( 'pantheon_cache_default_max_age' );
+	$custom = ! array_key_exists( $default_ttl, $options ) && ! $filtered;
+	$output = '';
```
less dense, but same thoughts re: bool variables as above.

pwtyler reviewed (approved):
I would love some cleanup and clarity as described in a couple places, but will leave at your discretion.

jazzsequence committed:
use boolean variable names
`$is_filtered` instead of `$filtered`
`$has_custom_ttl` instead of `$custom` and `$custom_ttl`
